runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,424 @@
+                          @Nullable final Timestamp timestamp,
+                          @Nonnull final K key) {
+        try {
+            txBegin(timestamp);
+            return ((Table<K, V, M>) getTable(tableName)).get(key);
+        } finally {
+            txEnd();
+        }
+    }
+
+    /**
+     * Gets the count of records in the table.
+     *
+     * @param tableName Table name.
+     * @return Count of records.
+     */
+    public int count(@Nonnull final String tableName) {
+        return count(tableName, null);
+    }
+
+    /**
+     * Gets the count of records in the table at a particular timestamp.
+     *
+     * @param tableName Table name.
+     * @param timestamp Timestamp to perform the query on.
+     * @return Count of records.
+     */
+    public int count(@Nonnull final String tableName,
+                     @Nullable final Timestamp timestamp) {
+        try {
+            txBegin(timestamp);
+            return getTable(tableName).count();
+        } finally {
+            txEnd();
+        }
+    }
+
+    @Nonnull
+    private <K extends Message, V extends Message, M extends Message>
+    Collection<CorfuRecord<V, M>> scanAndFilter(@Nonnull final String tableName,
+                                                @Nullable Timestamp timestamp,
+                                                @Nonnull final Predicate<CorfuRecord<V, M>> p) {
+        try {
+            txBegin(timestamp);
+            return ((Table<K, V, M>) getTable(tableName)).scanAndFilter(p);
+        } finally {
+            txEnd();
+        }
+    }
+
+    /**
+     * Query by a secondary index.
+     *
+     * @param tableName Table name.
+     * @param indexName Index name. In case of protobuf-defined secondary index it is the field name.
+     * @param indexKey  Key to query.
+     * @param <K>       Type of Key.
+     * @param <V>       Type of Value.
+     * @param <I>       Type of index/secondary key.
+     * @return Result of the query.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message, I extends Comparable<I>>
+    QueryResult<Entry<K, V>> getByIndex(@Nonnull final String tableName,
+                                        @Nonnull final String indexName,
+                                        @Nonnull final I indexKey) {
+        return new QueryResult<>(((Table<K, V, M>) getTable(tableName)).getByIndex(indexName, indexKey));
+    }
+
+    private <V extends Message, M extends Message, R>
+    Collection<R> initializeResultCollection(QueryOptions<V, M, R> queryOptions) {
+        if (!queryOptions.isDistinct()) {
+            return new ArrayList<>();
+        }
+        if (queryOptions.getComparator() != null) {
+            return new TreeSet<>(queryOptions.getComparator());
+        }
+        return new HashSet<>();
+    }
+
+    private <V extends Message, M extends Message, R>
+    Collection<R> transform(Collection<CorfuRecord<V, M>> queryResult,
```

Comment:
Can we use List or Set here or anything else? As we discussed using Collection is almost a bad idea